v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,497 @@
+                _tlgWriteTransfer_EventWriteTransfer(_tlgProv, &self_described_event->_tlgChannel, ((void*)0), ((void*)0), _tlgIdx, _tlgData);
+            }
+        }
+    }
+}
+
+static const char event_name_critical[] = "LogCritical";
+static const char event_name_error[] = "LogError";
+static const char event_name_warning[] = "LogWarning";
+static const char event_name_info[] = "LogInfo";
+static const char event_name_verbose[] = "LogVerbose";
+static const char event_name_unknown[] = "Unknown";
+
+static void log_sink_etw_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line, const char* message_format, ...)
+{
+    if (message_format == NULL)
+    {
+        /* Codes_SRS_LOG_SINK_ETW_01_001: [ If message_format is NULL, log_sink_etw.log_sink_log shall return. ]*/
+        (void)printf("Invalid arguments: LOG_LEVEL log_level=%" PRI_MU_ENUM ", LOG_CONTEXT_HANDLE log_context=%p, const char* file=%s, const char* func=%s, int line=%d, const char* message_format=%s\r\n",
+            MU_ENUM_VALUE(LOG_LEVEL, log_level), log_context, file, func, line, message_format);
+    }
+    else
+    {
+        const LOG_CONTEXT_PROPERTY_VALUE_PAIR* value_pairs;
+        uint16_t values_count;
+
+        internal_log_sink_etw_lazy_register_provider();
+
+        /* Codes_SRS_LOG_SINK_ETW_01_010: [ log_sink_etw_log shall emit a self described event that shall have the name of the event as follows: ]*/
+
+        if (log_context != NULL)
+        {
+            /* Codes_SRS_LOG_SINK_ETW_01_048: [ If log_context is not NULL: ]*/
+
+            /* Codes_SRS_LOG_SINK_ETW_01_049: [ log_sink_etw.log_sink_log shall call log_context_get_property_value_pair_count to obtain the count of properties that are to be added to the ETW event. ]*/
+            value_pairs = log_context_get_property_value_pairs(log_context);
+            /* Codes_SRS_LOG_SINK_ETW_01_050: [ log_sink_etw.log_sink_log shall call log_context_get_property_value_pairs to obtain the properties that are to be added to the ETW event. ]*/
+            values_count = (uint16_t)log_context_get_property_value_pair_count(log_context);
+        }
+        else
+        {
+            /* Codes_SRS_LOG_SINK_ETW_01_025: [ If log_context is NULL only the fields content, file, func and line shall be added to the ETW event. ]*/
+            value_pairs = NULL;
+            values_count = 0;
+        }
+
+        va_list args;
+        va_start(args, message_format);
+
+        switch (log_level)
+        {
+        default:
+            /* Codes_SRS_LOG_SINK_ETW_01_017: [ Otherwise the event name shall be Unknown. ]*/
+            /* Codes_SRS_LOG_SINK_ETW_01_024: [ Otherwise the ETW logging level shall be TRACE_LEVEL_NONE. ]*/
+            internal_emit_self_described_event(event_name_unknown, sizeof(event_name_unknown), TRACE_LEVEL_NONE, value_pairs, values_count, message_format, file, func, line, args);
```

Comment:
>values_count [](http://example.com/codeflow?start=126&length=12)

Last comment (almost):

Any chance the mighty "context" might provide some precomputed aggregated data (in general). So we don't have the serialize every time the same properties (get some "bulk" of them already in serialized form) with their numbers etc. Effectively context should do more in general, so not every `log_sink_etw_log` has to go through the same steps for the context.

Reply:
Yeah, was thinking of this too while writing the code. As in - maybe the context can be friendlier to tracelogging so that we can memcpy (or not event that).
I'll add a task for this.

Reply:
ID Work Item Type Title State Assigned To Effort Remaining Work Original Estimate Value Area Iteration Path Tags
22271413 Task Optimize the context layout to be friendlier to tracelogging To Do           One